# **Automatidata project**
**Course 2 - Get Started with Python**

Welcome to the Automatidata Project!

You have just started as a data professional in a fictional data consulting firm, Automatidata. Their client, the New York City Taxi and Limousine Commission (New York City TLC), has hired the Automatidata team for its reputation in helping their clients develop data-based solutions.

The team is still in the early stages of the project. Previously, you were asked to complete a project proposal by your supervisor, DeShawn Washington. You have received notice that your project proposal has been approved and that New York City TLC has given the Automatidata team access to their data. To get clear insights, New York TLC's data must be analyzed, key variables identified, and the dataset ensured it is ready for analysis.

A notebook was structured and prepared to help you in this project. Please complete the following questions.

# Course 2 End-of-course project: Inspect and analyze data

In this activity, you will examine data provided and prepare it for analysis.  This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>    

**The purpose** of this project is to investigate and understand the data provided.
  
**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings. 
<br/>  
*This activity has three parts:*

**Part 1:** Understand the situation 
* Prepare to understand and organize the provided taxi cab dataset and information.

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities.

* Compile summary information about the data to inform next steps.

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into specific variables.


<br/> 
Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work. 



# **Identify data types and relevant variables using Python**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## PACE: **Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided taxi cab information? 

To best prepare, I first reviewed the project scenario, team expectations, and the data dictionary describing each column. I also explored the dataset structure using tools like df.info(), df.describe(), and df.head() to understand column types, missing values, and potential anomalies. Organizing the data involved identifying which variables are relevant for predictive modeling, converting datetime fields to proper formats, and noting which columns require cleaning or transformation. This preparation ensures the dataset is structured, consistent, and ready for deeper analysis or modeling.

<img src="images/Analyze.png" width="100" height="100" align=left>

## PACE: **Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Build dataframe**
















Create a pandas dataframe for data learning, and future exploratory data analysis (EDA) and statistical activities.

**Code the following,**

*   import pandas as `pd`. pandas is used for buidling dataframes.

*   import numpy as `np`. numpy is imported with pandas

*   `df = pd.read_csv('Datasets\NYC taxi data.csv')`

**Note:** pair the data object name `df` with pandas functions to manipulate data, such as `df.groupby()`.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [1]:
#Import libraries and packages listed above
import pandas as pd
import numpy as np

# Load dataset into dataframe
df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')
print("done")

done


### **Task 2b. Understand the data - Inspect the data**

View and inspect summary information about the dataframe by coding the following:

1. `df.head(10)`
2. `df.info()`
3. `df.describe()`

Consider the following two questions:

**Question 1:** When reviewing the `df.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Question 2:** When reviewing the `df.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values?

Question 1: Most variables are numeric (int and float), and there are no null values. Some columns like the datetime fields are stored as strings(object), so they need conversion. All values are technically present, but some variables may still have invalid entries.

Question 2: 
-Negative fare_amount, total_amount, and extra, which shouldnt be possible
-Extremely high tip_amount of about $200 and total_amount is over $1200
-trip_distance has entries of 0 miles, which may indicate invalid or canceled trips
-RatecodeID has values up to 99 max, which is far outside the expected 1-6 range

In [2]:
df.head(10)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   
5    23345809         2   03/25/2017 8:34:11 PM   03/25/2017 8:42:11 PM   
6    37660487         2   05/03/2017 7:04:09 PM   05/03/2017 8:03:47 PM   
7    69059411         2   08/15/2017 5:41:06 PM   08/15/2017 6:03:05 PM   
8     8433159         2   02/04/2017 4:17:07 PM   02/04/2017 4:29:14 PM   
9    95294817         1   11/10/2017 3:20:29 PM   11/10/2017 3:40:55 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   
5                6           2.30           1                  N   
6                1          12.83           1                  N   
7                1           2.98           1                  N   
8                1           1.20           1                  N   
9                1           1.60           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   
5           161           236             1          9.0    0.5      0.5   
6            79           241             1         47.5    1.0      0.5   
7           237           114             1         16.0    1.0      0.5   
8           234           249             2          9.0    0.0      0.5   
9           239           237             1         13.0    0.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80  
5        2.06           0.0                    0.3         12.36  
6        9.86           0.0                    0.3         59.16  
7        1.78           0.0                    0.3         19.58  
8        0.00           0.0                    0.3          9.80  
9        2.75           0.0                    0.3         16.55

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [4]:
df.describe()

Unnamed: 0      VendorID  passenger_count  trip_distance  \
count  2.269900e+04  22699.000000     22699.000000   22699.000000   
mean   5.675849e+07      1.556236         1.642319       2.913313   
std    3.274493e+07      0.496838         1.285231       3.653171   
min    1.212700e+04      1.000000         0.000000       0.000000   
25%    2.852056e+07      1.000000         1.000000       0.990000   
50%    5.673150e+07      2.000000         1.000000       1.610000   
75%    8.537452e+07      2.000000         2.000000       3.060000   
max    1.134863e+08      2.000000         6.000000      33.960000   

         RatecodeID  PULocationID  DOLocationID  payment_type   fare_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean       1.043394    162.412353    161.527997      1.336887     13.026629   
std        0.708391     66.633373     70.139691      0.496211     13.243791   
min        1.000000      1.000000      1.000000      1.000000   -120.000000   
25%        1.000000    114.000000    112.000000      1.000000      6.500000   
50%        1.000000    162.000000    162.000000      1.000000      9.500000   
75%        1.000000    233.000000    233.000000      2.000000     14.500000   
max       99.000000    265.000000    265.000000      4.000000    999.990000   

              extra       mta_tax    tip_amount  tolls_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000   
mean       0.333275      0.497445      1.835781      0.312542   
std        0.463097      0.039465      2.800626      1.399212   
min       -1.000000     -0.500000      0.000000      0.000000   
25%        0.000000      0.500000      0.000000      0.000000   
50%        0.000000      0.500000      1.350000      0.000000   
75%        0.500000      0.500000      2.450000      0.000000   
max        4.500000      0.500000    200.000000     19.100000   

       improvement_surcharge  total_amount  
count           22699.000000  22699.000000  
mean                0.299551     16.310502  
std                 0.015673     16.097295  
min                -0.300000   -120.300000  
25%                 0.300000      8.750000  
50%                 0.300000     11.800000  
75%                 0.300000     17.800000  
max                 0.300000   1200.290000

### **Task 2c. Understand the data - Investigate the variables**

Sort and interpret the data table for two variables:`trip_distance` and `total_amount`.

**Answer the following three questions:**

**Question 1:** Sort your first variable (`trip_distance`) from maximum to minimum value, do the values seem normal?

**Question 2:** Sort by your second variable (`total_amount`), are any values unusual?

**Question 3:** Are the resulting rows similar for both sorts? Why or why not?

QUESTION 1:
Top 10 highest distances:
 -Range from ~28 to 34 miles
 -Corresponding fares seems realistic looking at $150 for ~34 miles 
 -So yes the values appear normal for long ride like JFK or Newark Airport trips
 
QUESTION 2:
Top 20 highest total amounts:
 -One row with fare = $999.99 and total = $1200.29 — clearly an outlier
 -Several trips have 0 trip distance but very high fares (e.g., $200 fare for 0 miles) — likely data entry errors or special service charges

Bottom 20 values:
 -Many negative total and fare amounts, like:
  -Fare = -120.00, total = -120.30
  -Total amounts of $0.00 or close to zero
 -These rows likely represent canceled trips, disputes, or data issues
 
QUESTION 3:
No, they are not similar:
 -trip_distance high values are valid long trips
 -total_amount high values include outliers, errors, and 0-distance rides with extreme charges

This tells us:
 -trip_distance is a more stable metric
 -total_amount is influenced by multiple factors (tips, tolls, errors)

In [7]:
# Sort the data by trip_distance from maximum to minimum
trip_distance_sorted = df.sort_values(by='trip_distance', ascending=False)
print("Top 10 Longest Trips (by trip_distance):")
print(trip_distance_sorted[['trip_distance', 'fare_amount', 'total_amount']].head(10))

Top 10 Longest Trips (by trip_distance):
       trip_distance  fare_amount  total_amount
9280           33.96       150.00        150.30
13861          33.92       200.01        258.21
6064           32.72       107.00        179.06
10291          31.95       131.00        131.80
29             30.83        80.00        111.38
18130          30.50        90.50        119.31
5792           30.33        52.00         73.20
15350          28.23        52.00         62.96
10302          28.20        52.00         70.27
2592           27.97        52.00         63.06


In [8]:
# Sort the data by total_amount from maximum to minimum
total_amount_top = df.sort_values(by='total_amount', ascending=False)
print("\nTop 20 Trips with Highest Total Amounts:")
print(total_amount_top[['trip_distance', 'fare_amount', 'total_amount']].head(20))


Top 20 Trips with Highest Total Amounts:
       trip_distance  fare_amount  total_amount
8476            2.60       999.99       1200.29
20312           0.00       450.00        450.30
13861          33.92       200.01        258.21
12511           0.00       175.00        233.74
15474           0.00       200.00        211.80
6064           32.72       107.00        179.06
16379          25.50       140.00        157.06
3582            7.30       152.00        152.30
11269           0.00       120.00        151.82
9280           33.96       150.00        150.30
1928           12.50       120.00        137.80
10291          31.95       131.00        131.80
6708            0.32       100.00        126.00
11608          23.00        99.50        123.30
908            26.12       100.00        121.56
7281            0.00       100.00        120.96
18130          30.50        90.50        119.31
13621          19.80       105.00        115.94
13359           0.00        75.00        111.9

In [9]:
# Sort the data by total amount and print the bottom 20 values
total_amount_bottom = df.sort_values(by='total_amount', ascending=True)
print("\nBottom 20 Trips with Lowest Total Amounts:")
print(total_amount_bottom[['trip_distance', 'fare_amount', 'total_amount']].head(20))


Bottom 20 Trips with Lowest Total Amounts:
       trip_distance  fare_amount  total_amount
12944           0.17      -120.00       -120.30
20698           0.70        -4.50         -5.80
17602           0.46        -4.00         -5.80
11204           0.64        -4.50         -5.30
14714           0.40        -4.00         -5.30
8204            0.41        -3.50         -4.80
20317           0.24        -3.50         -4.80
10281           0.00        -2.50         -4.30
5448            0.25        -3.50         -4.30
4423            0.06        -3.00         -4.30
18565           0.10        -3.00         -3.80
314             0.12        -2.50         -3.80
5758            0.02        -2.50         -3.80
1646            0.04        -2.50         -3.30
10506           0.00         0.00          0.00
4402            7.06         0.00          0.00
5722            0.00         0.00          0.00
22566           0.00         0.00          0.00
19067           0.10         0.00          0

In [12]:
# How many of each payment type are represented in the data?
payment_type_labels = {
    1: "Credit card",
    2: "Cash",
    3: "No charge",
    4: "Dispute",
    5: "Unknown",
    6: "Voided trip"
}

# Replace codes with labels for display purposes (doesn't change original DataFrame)
payment_type_named = df['payment_type'].map(payment_type_labels)
print("\nPayment type counts (with labels):")
print(payment_type_named.value_counts())


Payment type counts (with labels):
Credit card    15265
Cash            7267
No charge        121
Dispute           46
Name: payment_type, dtype: int64


According to the data dictionary, the payment method was encoded as follows:

1 = Credit card  
2 = Cash  
3 = No charge  
4 = Dispute  
5 = Unknown  
6 = Voided trip

In [14]:
# What is the average tip for trips paid for with credit card?
avg_tip_credit = df[df['payment_type'] == 1]['tip_amount'].mean()
print(f"Average tip for credit card payments: ${avg_tip_credit:.2f}")

# What is the average tip for trips paid for with cash?
avg_tip_cash = df[df['payment_type'] == 2]['tip_amount'].mean()
print(f"Average tip for cash payments: ${avg_tip_cash:.2f}")

Average tip for credit card payments: $2.73
Average tip for cash payments: $0.00


In [15]:
# How many times is each vendor ID represented in the data?
vendor_counts = df['VendorID'].value_counts()
print("\nNumber of rides by Vendor ID:")
print(vendor_counts)


Number of rides by Vendor ID:
2    12626
1    10073
Name: VendorID, dtype: int64


In [16]:
# What is the mean total amount for each vendor?
vendor_total_avg = df.groupby('VendorID')['total_amount'].mean()
print("\nAverage total amount by Vendor ID:")
print(vendor_total_avg)


Average total amount by Vendor ID:
VendorID
1    16.298119
2    16.320382
Name: total_amount, dtype: float64


In [22]:
# Filter the data for credit card payments only
credit_card_only = df[df['payment_type'] == 1]

# Filter the credit-card-only data for passenger count only
passenger_count_distribution = credit_card_only['passenger_count'].value_counts().sort_index()
print("Passenger count distribution for credit card payments:")
print(passenger_count_distribution)

Passenger count distribution for credit card payments:
0       27
1    10977
2     2168
3      600
4      267
5      775
6      451
Name: passenger_count, dtype: int64


In [23]:
# Calculate the average tip amount for each passenger count(credit card payments only)
avg_tip_by_passenger = credit_card_only.groupby('passenger_count')['tip_amount'].mean()
print("\nAverage tip by passenger count (Credit Card only):")
print(avg_tip_by_passenger)


Average tip by passenger count (Credit Card only):
passenger_count
0    2.610370
1    2.714681
2    2.829949
3    2.726800
4    2.607753
5    2.762645
6    2.643326
Name: tip_amount, dtype: float64


<img src="images/Construct.png" width="100" height="100" align=left>

## PACE: **Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project. 




<img src="images/Execute.png" width="100" height="100" align=left>

## PACE: **Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response.


### **Given your efforts, what can you summarize for DeShawn and the data team?**

*Note for Learners: Your notebook should contain data that can address Luana's requests. Which two variables are most helpful for building a predictive model for the client: NYC TLC?*

Recommended Variables:
-trip_distance
  -Strong correlation to fare amount and total amount.
  -As distance increases, fare logically increases.
  -Clean, continuous numeric data with few outliers.

-payment_type
  -Impacts tip inclusion (e.g., credit cards track tips).
  -Helps distinguish cash vs. card payment behavior.
  -Can explain variations in tip_amount and total_amount.



**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.